instance_control/create_instance.py

```diff
@@ -0,0 +1,480 @@
+            #
+            # This has to be our root EBS.
+            #
+            # If the Encrypted flag is present, we have to delete
+            # it even if it matches the actual snapshot setting,
+            # otherwise amazon will complain rather loudly.
+            #
+            # Take a deep copy before deleting the key:
+            #
+            bd = copy.deepcopy(bd)
+
+            root_ebs = bd['Ebs']
+            if 'Encrypted' in root_ebs:
+                del (root_ebs['Encrypted'])
+
+            block_devices.append(bd)
+        else:
+            # ignore any ephemeral volumes (aka. instance storage)
+            block_devices.append({'DeviceName': bd['DeviceName'],
+                                  'NoDevice': ''})
+
+    volume_name = 'kafka-logs-ebs'
+    if options['create_ebs']:
+        create_tagged_volume(ec2, options, subnet['AvailabilityZone'], volume_name)
+
+    user_data = options['user_data']
+    user_data['volumes']['ebs']['/dev/xvdk'] = volume_name
+    taupage_user_data = '#taupage-ami-config\n{}'.format(yaml.safe_dump(user_data))
+
+    resp = ec2.run_instances(
+        ImageId=ami.id,
+        MinCount=1,
+        MaxCount=1,
+        SecurityGroupIds=[security_group_id],
+        UserData=taupage_user_data,
+        InstanceType=options['instance_type'],
+        SubnetId=subnet['SubnetId'],
+        PrivateIpAddress=ip['PrivateIp'],
+        BlockDeviceMappings=block_devices,
+        IamInstanceProfile={'Arn': options['instance_profile']['Arn']},
+        DisableApiTermination=False)
+
+    instance = resp['Instances'][0]
+    instance_id = instance['InstanceId']
+
+    ec2.create_tags(Resources=[instance_id],
+                    Tags=[{'Key': 'Name', 'Value': options['cluster_name']}])
+
+    # wait for instance to initialize before we can assign a
+    # public IP address to it or tag the attached volume
+    while True:
+        resp = ec2.describe_instances(InstanceIds=[instance_id])
+        instance = resp['Reservations'][0]['Instances'][0]
+        if instance['State']['Name'] != 'pending':
+            break
+        time.sleep(5)
+
+    # add an auto-recovery alarm for this instance
+    cw = boto3.client('cloudwatch', region_name=region)
+    alarm_actions = ['arn:aws:automate:{}:ec2:recover'.format(region)]
+    if options['alarm_topics']:
+        alarm_actions.append(options['alarm_topics'][region])
+
+    cw.put_metric_alarm(AlarmName='{}-{}-auto-recover'.format(options['cluster_name'], instance_id),
+                        AlarmActions=alarm_actions,
+                        MetricName='StatusCheckFailed_System',
+                        Namespace='AWS/EC2',
+                        Statistic='Minimum',
+                        Dimensions=[{
+                            'Name': 'InstanceId',
+                            'Value': instance_id
+                        }],
+                        Period=60,  # 1 minute
+                        EvaluationPeriods=2,
```

Comment:
Does it mean that we won't be able to perform some manual maintenance longer than for 2 minutes?